NDTensors/ext/NDTensorsCUDAExt/mul.jl

```diff
@@ -0,0 +1,45 @@
+# This was calling generic matrix multiplication.
+# TODO: Raise an issue with `CUDA.jl`.
+function LinearAlgebra.mul!(
+  CM::Exposed{<:CuArray,<:LinearAlgebra.Transpose},
+  AM::Exposed{<:CuArray},
+  BM::Exposed{<:CuArray},
+  α,
+  β,
+)
+  mul!(transpose(CM), transpose(BM), transpose(AM), α, β)
+  return unexpose(CM)
+end
+
+# This was calling generic matrix multiplication.
+# TODO: Raise an issue with `CUDA.jl`.
+function LinearAlgebra.mul!(
+  CM::Exposed{<:CuArray,<:LinearAlgebra.Adjoint},
+  AM::Exposed{<:CuArray},
+  BM::Exposed{<:CuArray},
+  α,
+  β,
+)
+  mul!(CM', BM', AM', α, β)
+  return unexpose(CM)
+end
+
+## TODO I wasn't sure the best route to go here, if there is a better route than
+## copy please let me know!
```

Comment:
Copying seems reasonable, maybe there is something more clever but with complicated wrapper types that may be the best we can do (without flattening the wrappers using `StridedView`, but as you know that will require more effort to get working for GPU backends). Is there a test for this case?

Reply:
@mtfishman Yes there is a test case for this its in the `diag.jl` test - these lines https://github.com/ITensor/ITensors.jl/blob/d9265a22773e4948c38cec28a39e1489bf89b90c/NDTensors/test/diag.jl#L41C1-L42C48

Reply:
I think it would be good to add a test directly for `mul!` so it isn't just being tested indirectly through `contract`.

Reply:
Do you have a preference on where the tests should live? I.e. is it fine to add lines to `diag.jl` which derive a `mul!` test from the existing code? Something like 
```
S1M = dev(Matrix{elt}(undef, (2,12)))
SM = array(S)
VM = transpose(reshape(array(V), (12,2)))
NDTensors.mul!!(S1M, SM, VM, 1.0, 0.0)
@test @allowscalar reshape(S1M, (2,3,4)) ≈ array(S1)
```

Reply:
I think it is better to make a more minimal self-contained test in [Unwrap/test/runtests.jl](https://github.com/ITensor/ITensors.jl/pull/1245/files#diff-41118dc0c84433e17fb9c7eae2e5734daff5e01501d8d182050f108233600f6b) that is independent of `Tensor`/`TensorStorage` objects. For example, you could use something like `transpose(reshape(randn(2, 2)', 2, 2))` to make a wrapper with that type. That way we can better track what issues there are in different backends, and they won't be sensitive to changes we make to the design of the `Tensor` type.